crates/noirc_evaluator/src/ssa/acir_gen/operations/store.rs

```diff
@@ -1,34 +1,66 @@
+use acvm::FieldElement;
+
 use crate::{
     ssa::{
-        acir_gen::{internal_var_cache::InternalVarCache, memory_map::MemoryMap, InternalVar},
+        acir_gen::{
+            constraints::{add, mul_with_witness, subtract},
+            internal_var_cache::InternalVarCache,
+            memory_map::MemoryMap,
+            InternalVar,
+        },
         context::SsaContext,
-        mem::{self, ArrayId},
-        node::NodeId,
+        mem::{self},
+        node::Operation,
     },
     Evaluator,
 };
 
 pub(crate) fn evaluate(
-    array_id: ArrayId,
-    index: NodeId,
-    value: NodeId,
+    store: &Operation,
     memory_map: &mut MemoryMap,
     var_cache: &mut InternalVarCache,
     evaluator: &mut Evaluator,
     ctx: &SsaContext,
 ) -> Option<InternalVar> {
-    //maps the address to the rhs if address is known at compile time
-    let index = var_cache.get_or_compute_internal_var_unwrap(index, evaluator, ctx);
-    let value = var_cache.get_or_compute_internal_var_unwrap(value, evaluator, ctx);
+    if let Operation::Store { array_id, index, value, predicate } = *store {
+        //maps the address to the rhs if address is known at compile time
+        let index = var_cache.get_or_compute_internal_var_unwrap(index, evaluator, ctx);
+        let value = var_cache.get_or_compute_internal_var_unwrap(value, evaluator, ctx);
+
+        match index.to_const() {
+            Some(index) => {
+                let idx = mem::Memory::as_u32(index);
+                let absolute_adr = ctx.mem[array_id].absolute_adr(idx);
+                let value_with_predicate = if let Some(predicate) = predicate {
+                    if predicate.is_dummy() || ctx.is_one(predicate) {
+                        value
+                    } else if ctx.is_zero(predicate) {
+                        return None;
+                    } else {
+                        let pred =
+                            var_cache.get_or_compute_internal_var_unwrap(predicate, evaluator, ctx);
+                        let dummy = memory_map
+                            .load_array_element_constant_index(&ctx.mem[array_id], idx)
+                            .unwrap();
+                        let x = mul_with_witness(evaluator, pred.expression(), value.expression());
+                        let y = subtract(
+                            dummy.expression(),
```

Comment:
Can you explain why the variable is called dummy and what `dummy.expression()` is supposed to output?

Reply:
The `dummy` vs `None` conflation is something we should handle in a separate PR

Reply:
It's on the line above, I have renamed it to dummy_load for more clarity.